src/free_addresses/free_addresses.go

```diff
@@ -0,0 +1,181 @@
+	prev := ip[start]
+	ip[start] += add
+	if ip[start] < prev {
+		for j := start - 1; j >= 0; j-- {
+			ip[j]++
+			if ip[j] > 0 {
+				break
+			}
+		}
+	}
+}
+
+func scanSubNetwork(subNetwork string, exe IExecute) ([]strfmt.IPv4, string, int) {
+	ip, cidr, err := net.ParseCIDR(subNetwork)
+	if err != nil {
+		wrapped := errors.Wrapf(err, "Network %s is not a valid CIDR", subNetwork)
+		logrus.WithError(wrapped).Warn("ParseCIDR")
+		return nil, wrapped.Error(), -1
+	}
+	if cidr.String() != subNetwork {
+		stderr := fmt.Sprintf("Requested CIDR %s is not equal to provided subNetwork %s", cidr.String(), subNetwork)
+		logrus.Warn(stderr)
+		return nil, stderr, -1
+	}
+	o, e, exitCode := exe.Execute("nmap", "-sn", "-n", "-oX", "-", "--max-retries", "0", subNetwork)
+	if exitCode != 0 {
+		logrus.Warnf("nmap failed with exit-code %d: %s", exitCode, e)
+		return nil, e, exitCode
+	}
+	var nmaprun nmaprun
+	err = xml.Unmarshal([]byte(o), &nmaprun)
+	if err != nil {
+		logrus.WithError(err).Warn("XML Unmarshal")
+		return nil, err.Error(), -1
+	}
+	occupiedAddesses := make(map[string]int)
+	for _, h := range nmaprun.Hosts {
+		if h.Status.State == "up" {
+			for _, a := range h.Addresses {
+				if a.AddrType == "ipv4" {
+					occupiedAddesses[a.Addr] = 0
+					break
+				}
+			}
+		}
+	}
+	ret := make([]strfmt.IPv4, 0)
+	for ip = ip.To4(); cidr.Contains(ip); inc(ip) {
+		_, ok := occupiedAddesses[ip.String()]
+		if !ok {
+			ret = append(ret, strfmt.IPv4(ip.String()))
+		}
+	}
+	return ret, "", 0
+}
+
+func scanNetwork(network string, exe IExecute) (*models.FreeNetworkAddresses, string, int) {
+	ret := models.FreeNetworkAddresses{
+		Network: network,
+	}
+	ip, cidr, err := net.ParseCIDR(network)
+	if err != nil {
+		wrapped := errors.Wrapf(err, "Network %s is not a valid CIDR", network)
+		logrus.WithError(wrapped).Warn("ParseCIDR")
+		return &ret, wrapped.Error(), -1
+	}
+	if cidr.String() != network {
+		stderr := fmt.Sprintf("Requested CIDR %s is not equal to provided network %s", cidr.String(), network)
+		logrus.Warn(stderr)
+		return &ret, stderr, -1
+	}
+	ones, _ := cidr.Mask.Size()
+
+	var mask int
+
+	if ones < MinSubnetMaskSize {
+		mask = MinSubnetMaskSize
+	} else {
+		mask = ones
+	}
+	for ip = ip.To4(); cidr.Contains(ip) && len(ret.FreeAddresses) < AddressLimit; incSubnet(ip, mask) {
+		subNetwork := fmt.Sprintf("%s/%d", ip.To4().String(), mask)
+		result, e, exitCode := scanSubNetwork(subNetwork, exe)
+		if exitCode != 0 {
+			return &ret, e, exitCode
+		}
+		ret.FreeAddresses = append(ret.FreeAddresses, result...)
+	}
+	return &ret, "", 0
+}
+
+func GetFreeAddresses(freeAddressesRequestStr string, e IExecute) (stdout string, stderr string, exitCode int) {
+	var freeAddressesRequest models.FreeAddressesRequest
+	err := json.Unmarshal([]byte(freeAddressesRequestStr), &freeAddressesRequest)
+	if err != nil {
+		logrus.WithError(err).Errorf("FreeAddresses: json.Unmarshal")
+		return "", err.Error(), -1
```

Comment:
Why not add the error message to the error that is created here?

Reply:
It runs in a process.  The return values are stdout, stderr, exitCode